### ISSUE TITLE
typevar solver can't solve a generic union

### ISSUE THREAD

**Issue by suleymanozkeskin** [astral-sh/ty#1703]
### Summary

# Summary

Incomplete type narrowing with `TypeIs` on generic types after early return

## Description

Type narrowing fails for **generic types** when using a `TypeIs` type guard with an early return pattern. Pyright correctly narrows the type in this scenario, but `ty` does not. The issue does **not** occur with simple (non-generic) classes.

## Minimal Reproducible Example

```python
# exp.py
from typing import Union, Generic, TypeVar
from typing_extensions import TypeIs

T = TypeVar('T', covariant=True)
E = TypeVar('E', covariant=True)


class Ok(Generic[T]):
    def __init__(self, value: T):
        self._value = value
    
    @property
    def ok_value(self) -> T:
        return self._value
    
    def is_err(self) -> bool:
        return False


class Err(Generic[E]):
    def __init__(self, error: E):
        self._error = error
    
    @property
    def err_value(self) -> E:
        return self._error
    
    def is_err(self) -> bool:
        return True


Result = Union[Ok[T], Err[E]]


def is_err(result: Result[T, E]) -> TypeIs[Err[E]]:
    """Type guard to check if a result is an Err"""
    return result.is_err()

# ------------------------------------------------------------

def some_function() -> Result[int, Exception]:
    condition = False
    if condition is False:
        return Err(Exception("Test error"))
    return Ok(1)

def use_some_function():    
    result = some_function()
    
    if is_err(result):
        print(result.err_value)
        return None

    # After the is_err() check and early return,
    # result should be narrowed to Ok[int]
    okay = result.ok_value  # <- ty reports warning here
    print(okay)
```

## Expected Behavior

After the `if is_err(result): return None` check, `result` should be narrowed to `Ok[int]`, making `result.ok_value` valid without warnings.

## Actual Behavior

```bash
uv run ty check exp.py
warning[possibly-missing-attribute]: Attribute `ok_value` may be missing on object of type `(Ok[int] & ~Err[Unknown]) | (Err[Exception] & ~Err[Unknown])`
  --> exp.py:57:12
   |
55 |     # After the is_err() check and early return,
56 |     # result should be narrowed to Ok[int]
57 |     okay = result.ok_value  # <- ty reports warning here
   |            ^^^^^^^^^^^^^^^
58 |     print(okay)
   |
info: rule `possibly-missing-attribute` is enabled by default
```

## Comparison with Pyright

Pyright correctly handles this pattern with 0 errors, 0 warnings for both generic and non-generic types.

- Simple classes with `TypeIs`: Works correctly in both `ty` and Pyright
- Generic classes with `TypeIs`: Fails in `ty`, works in Pyright

## Notes

- Pattern matching (`match`/`case`) works correctly in both type checkers, even with generics
- The issue specifically occurs when combining:
  1. Generic types (e.g., `Union[Ok[T], Err[E]]`)
  2. `TypeIs` type guards
  3. Early return control flow
- Type narrowing appears to be incomplete when the type guard eliminates one branch of a generic union

## Environment

- `ty` version: 0.0.1a29
- Python version: 3.13

### Version

0.0.1a29

**Comment by carljm** [comment]
Thanks for the report! This looks like a bug indeed. I think it is a limitation of our current typevar solver; it isn't able to match the generic union parameter type against the union argument and solve both typevars. [This simplified example](https://play.ty.dev/32048e84-3f13-4553-aac7-5d51daccaaa2) (without `TypeIs`, and also without the generic implicit alias, to clarify that's not part of the problem either) shows the root cause:

```py
from typing import Union, Generic, TypeVar

T = TypeVar('T', covariant=True)
E = TypeVar('E', covariant=True)


class Ok(Generic[T]):
    def __init__(self, value: T):
        self._value = value
    
    @property
    def ok_value(self) -> T:
        return self._value
    
    def is_err(self) -> bool:
        return False


class Err(Generic[E]):
    def __init__(self, error: E):
        self._error = error
    
    @property
    def err_value(self) -> E:
        return self._error
    
    def is_err(self) -> bool:
        return True



def err_kind(result: Union[Ok[T], Err[E]]) -> E | None:
    pass

# ------------------------------------------------------------

def f(result: Union[Ok[int], Err[Exception]]):
    reveal_type(err_kind(result))  # reveals `Unknown | None`, should reveal `Exception | None`
```

**Comment by carljm** [comment]
Even smaller repro:

```py
class A[T]:
    pass

class B[T]:
    pass

def f[T](x: A[T] | B[T]) -> A[T]: ...

def _(x: A[int] | B[int]):
    reveal_type(f(x)) # reveals A[Unknown], should be A[int]
```

**Comment by dhnam** [comment]
I believe this belongs here?
```py
from typing import reveal_type
class Test[A]:
    def __init__(self, test: A | list[A]):
        self._test:A | list[A] = test

a: Test[int] = Test([1]) # error
b: Test[int] = Test(1) # okay

reveal_type(Test([1])) # revealed: Test[list[int] | int], should: Test[int]
```

**Comment by carljm** [comment]
@dhnam I think that is more precisely #2870 , though they are related issues.

**Comment by dcreager** [comment]
Possibly related to and/or a dup of #3557

**Comment by jelle-openai** [comment]
Here's another practical example that appears to be the same issue:

```python
from collections.abc import Iterator
from typing import AnyStr, Generic, TypeAlias


class PathLike(Generic[AnyStr]): ...
class Path(PathLike[str]): ...

GenericPath: TypeAlias = AnyStr | PathLike[AnyStr]


def walk(top: GenericPath[AnyStr]) -> Iterator[tuple[AnyStr, list[AnyStr], list[AnyStr]]]: ...


def all_files(root: str | Path) -> None:
    result = walk(root)
    for _, _, filenames in result:
        filenames.append(1)
```

(This is `os.walk`, but inlined so as not to depend on typeshed.)

| Expression | mypy | pyright | ty |
| --- | --- | --- | --- |
| `walk(root)` | `Iterator[tuple[str, list[str], list[str]]]` | `Iterator[tuple[str, list[str], list[str]]]` | `Iterator[tuple[Unknown, list[Unknown], list[Unknown]]]` |
| `filenames.append(1)` | error | error | no error |
